_posts/2022-01-24-publishing-typescript-npm-modules.md

```diff
@@ -0,0 +1,193 @@
+    "@mui/material": "^5.0.0",
+    "@emotion/styled": "^11.3.0"
+}
+```
+
+You might not have expected that `@emotion/styled` would show up in this list; we have to include it since its usually an optional dependency of `@mui/material` but it is required for generating Material UI themes.
+
+## Publishing to a Registry
+
+We are almost ready to publish our module, we just need to tell NPM where our compiled code is sitting which can be done by adding the following block to your `package.json`:
+
+```json
+{
+	"main": "dist/index.js",
+	"types": "dist/index.d.ts",
+	"files": ["/dist"]
+}
+```
+
+And finally, we can run `npm publish`.
+
+But wait! What's this error?!
+
+```sh
+npm ERR! code ENEEDAUTH
+npm ERR! need auth This command requires you to be logged in.
+npm ERR! need auth You need to authorize this machine using `npm adduser`
+```
+
+You need an account on NPM (with a verified email address) to be able to publish your package which you can [sign up for here](https://www.npmjs.com/signup). We can then run `npm adduser` with the user credentials we just created to sign in to the NPM registry from the command line.
+
+Let's try running `npm publish` again.
+
+And we should get the output...
+
+```sh
+npm notice
+npm notice 📦  tim-veletta-theme@1.0.0
+npm notice === Tarball Contents ===
+npm notice 81B  dist/index.d.ts
+npm notice 340B dist/index.js
+npm notice 489B package.json
+npm notice === Tarball Details ===
+npm notice name:          tim-veletta-theme
+npm notice version:       1.0.0
+npm notice filename:      tim-veletta-theme-1.0.0.tgz
+npm notice package size:  616 B
+npm notice unpacked size: 910 B
+npm notice shasum:        8d6517abb56fab745484a03e99b8351eb9137c19
+npm notice integrity:     sha512-gWgTDeJqBEfkd[...]IJ0aRjxpWnPDg==
+npm notice total files:   3
+npm notice
++ tim-veletta-theme@1.0.0
+```
+
+There it is! Your module has been added to the NPM registry and anyone using NPM can install it into their project. You can even search the [NPM registry](https://www.npmjs.com/) to see the details of your module.
+
+## Using the Package
+
+Let's test out the module in a new React project by installing it with:
+
+```sh
+npm install tim-veletta-theme
+```
+
+This should also install the `@mui/material` and `@emotion/styled` packages for us as they were listed as `peerDependencies` of our module. From here, we can inject our Material UI theme and see it in action!
+
+```jsx
+import { ThemeProvider, Typography } from '@mui/material';
+import theme from 'tim-veletta-theme';
+
+function App() {
+	return (
+		<ThemeProvider theme={theme}>
+			<Typography variant="h1" color="primary">
+				This should be in the primary color #40bfb4
+			</Typography>
+			<Typography variant="h1" color="secondary">
+				This should be in the secondary color #e15554
+			</Typography>
+		</ThemeProvider>
+	);
+}
+
+export default App;
+```
+
+When we run it in the browser, we should see the text in the primary and secondary colours we set in our module.
```

Comment:
You should probably include the package.json file in its entirety.

Reply:
All the package.json really shows is that `"tim-veletta-theme": "^1.0.1",` is installed, if you were to look through the package-lock.json it would show that @mui/material and @emotion/styled are also installed but I'm not sure of the need to include the file as I think it detracts from the readability.

Reply:
You've spent a bit of time building up an example, so it really helps to have it in it's entirety.

Nothing pisses me off more in a tech blog post then trying an example out but it doesn't work, and I don't have a complete example to diff against.

Reply:
Yeah thats fair, I've added in the 2 repositories that I used as an example, do you think thats enough or would you still prefer to see the package.json?

I'm just really hesitant to post all of this because I think its too much for what I'm trying to explain.

```
{
	"name": "module-test",
	"version": "0.1.0",
	"private": true,
	"dependencies": {
		"@testing-library/jest-dom": "^5.16.1",
		"@testing-library/react": "^12.1.2",
		"@testing-library/user-event": "^13.5.0",
		"react": "^17.0.2",
		"react-dom": "^17.0.2",
		"react-scripts": "5.0.0",
		"tim-veletta-theme": "^1.0.1",
		"web-vitals": "^2.1.4"
	},
	"scripts": {
		"start": "react-scripts start",
		"build": "react-scripts build",
		"test": "react-scripts test",
		"eject": "react-scripts eject"
	},
	"eslintConfig": {
		"extends": [
			"react-app",
			"react-app/jest"
		]
	},
	"browserslist": {
		"production": [
			">0.2%",
			"not dead",
			"not op_mini all"
		],
		"development": [
			"last 1 chrome version",
			"last 1 firefox version",
			"last 1 safari version"
		]
	}
}
```